mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/DeleteMessageListener.java

```diff
@@ -0,0 +1,145 @@
+                                 CassandraAttachmentDAOV2 attachmentDAO, CassandraAttachmentOwnerDAO ownerDAO,
+                                 CassandraAttachmentMessageIdDAO attachmentMessageIdDAO) {
+        this.imapUidDAO = imapUidDAO;
+        this.messageIdDAO = messageIdDAO;
+        this.messageDAO = messageDAO;
+        this.attachmentDAO = attachmentDAO;
+        this.ownerDAO = ownerDAO;
+        this.attachmentMessageIdDAO = attachmentMessageIdDAO;
+    }
+
+    private final CassandraMessageIdToImapUidDAO imapUidDAO;
+    private final CassandraMessageIdDAO messageIdDAO;
+    private final CassandraMessageDAO messageDAO;
+    private final CassandraAttachmentDAOV2 attachmentDAO;
+    private final CassandraAttachmentOwnerDAO ownerDAO;
+    private final CassandraAttachmentMessageIdDAO attachmentMessageIdDAO;
+
+    @Override
+    public Group getDefaultGroup() {
+        return new DeleteMessageListenerGroup();
+    }
+
+    @Override
+    public boolean isHandling(Event event) {
+        return event instanceof Expunged || event instanceof MailboxDeletion;
+    }
+
+    @Override
+    public void event(Event event) {
+        if (event instanceof Expunged) {
+            Expunged expunged = (Expunged) event;
+
+            Flux.fromIterable(expunged.getExpunged()
+                .values())
+                .map(MessageMetaData::getMessageId)
+                .map(CassandraMessageId.class::cast)
+                .concatMap(this::handleDeletion)
+                .then()
+                .block();
+        }
+        if (event instanceof MailboxDeletion) {
+            MailboxDeletion mailboxDeletion = (MailboxDeletion) event;
+
+            CassandraId mailboxId = (CassandraId) mailboxDeletion.getMailboxId();
+
+            messageIdDAO.retrieveMessages(mailboxId, MessageRange.all())
+                .map(ComposedMessageIdWithMetaData::getComposedMessageId)
+                .concatMap(metadata -> imapUidDAO.delete((CassandraMessageId) metadata.getMessageId(), mailboxId)
+                    .then(messageIdDAO.delete(mailboxId, metadata.getUid()))
+                    .then(handleDeletion((CassandraMessageId) metadata.getMessageId())))
+                .then()
+                .block();
+        }
+    }
+
+    private Mono<Void> handleDeletion(CassandraMessageId messageId) {
+        return Mono.just(messageId)
+            .filterWhen(this::isReferenced)
+            .flatMap(id -> readMessage(id)
+                .flatMap(this::deleteUnreferencedAttachments)
+                .then(messageDAO.delete(messageId)));
+    }
+
+    private Mono<MessageRepresentation> readMessage(CassandraMessageId id) {
+        return messageDAO.retrieveMessage(id, MessageMapper.FetchType.Metadata);
+    }
+
+    private Mono<Void> deleteUnreferencedAttachments(MessageRepresentation message) {
+        return Flux.fromIterable(message.getAttachments())
```

Comment:
are we actually reusing attachments between messages? can you point me to the code that does this?

Reply:
> are we actually reusing attachments between messages? 

No.

However I tended to avoid coupling and be double safe.

I'm OK to assume that it's not the case.

Reply:
You tell that because as you have an API to retrieve Attachment from Ids, you can't be sure how/where it's persisted?
AFAIK, it's almost impossible to have a perfect tracking strategy but having a class that reacts to events of the system and updates its projection (what we do in our Event Sourcing impl).
We probably will have to leave with such brittleness for the time being, right?

Reply:
> You tell that because as you have an API to retrieve Attachment from Ids, you can't be sure how/where it's persisted?

That's what I meant by coupling.

> We probably will have to leave with such brittleness for the time being, right?

Right.

However, https://github.com/linagora/james-project/pull/3061 solves this by enforcing the AttachmentManager to allocate AttachementId. With that work merged we will be able to safely infer there is not id reuse IMO.